### ISSUE TITLE
How to diagnose NaN predictions?

### ISSUE THREAD

**Issue by nchaly** [Azure/mmlspark#878]
Hi, I'm trying to understand the root cause for NaN predictions.

Using LightGBMRegressor-trained model in Databricks cluster, I sometimes receive NaNs as predictions.

My progress so far:

I checked trained model dump and found several leaves with "-nan" as value:
```
"left_child":{
  "leaf_index":4,
  "leaf_value": -nan,
  "leaf_weight":34758,
  "leaf_count":34758
}
```
```
"right_child":{
  "leaf_index":27,
  "leaf_value":-nan,
  "leaf_weight":4062,
  "leaf_count":4062
}
```

(I've reloaded model dump using local python lightgbm 2.3.1, and result is the same. So this must be strictly training-related).

I've tried to repartition dataframe in a couple of different ways, and nans appear for some items in a seemingly random fashion.  

**My primary question**: Are there known conditions when a leaf can have 'nan' as value, like value by default?

Second: If this is specific to partitioning, is there a "correct" way to do it? 

**Additional info:**

Using library com.microsoft.ml.spark:mmlspark_2.11:1.0.0-rc1

Estimated train size is about 10M, test set about 500K samples.

Environment:
Azure Databricks, Apache Spark 2.4.5, Scala 2.11

**Comment by imatiach-msft** [comment]
@nchaly Sorry about the issue you are having.  This seems like an issue with LightGBM, probably not related to the distributed code specifically - you might have more luck adding the issue here: https://github.com/Microsoft/LightGBM.  Can you try the latest build, which uses a newer version of LightGBM:
```
Maven Coordinates
com.microsoft.ml.spark:mmlspark_2.11:1.0.0-rc1-96-fce3c952-SNAPSHOT
Maven Resolver
https://mmlspark.azureedge.net/maven
```
Hopefully this might fix your issue?

**Comment by nchaly** [comment]
Thank you. 
Just ran an experiment with `com.microsoft.ml.spark:mmlspark_2.11:1.0.0-rc1-96-fce3c952-SNAPSHOT` - unfortunately NANs are still there.

**Comment by imatiach-msft** [comment]
@nchaly any way I could get a repro for this issue?  Although I think it is an issue in LightGBM codebase, I can take a look to see if I can fix it there.  If you can send a repro with a dataset + code, that would help both the LightGBM devs and me take a look into this.  Also, did you add the issue in https://github.com/Microsoft/LightGBM/issues?

**Comment by nchaly** [comment]
@imatiach-msft thank you.
The data I'm using is confidential, so I'd need some time to figure out on how to reproduce the issue.

**Comment by eromoe** [comment]
I also get all nan predictions . @nchaly Do you have a workaround now?

**Comment by F4RZ4D** [comment]
I have a similar issue, with 2.11:1.0.0-rc3. Did anybody get any insight on this issue?

Thanks.

**Comment by nchaly** [comment]
@imatiach-msft Sorry it's been some time. In general, I've been observing it all the time since I've submitted the issue. It happens more in situations when dataset size somewhat decreased. For example, if I split dataset to two parts and train separately - there would be NaNs than with full dataset. The smaller dataset - the more probability of NaNs. This happened with both rc1 and rc3.

As I use confidential datasets, it would take some time for me to obfuscate dataset and test it.
But I'm pretty sure - the issue is not specific to the data I use.
